### PR TITLE
Models with pre made grids

### LINKED ISSUE #183: models with pre-made grids
When training a new model, we often want to tune its hyperparameters; however, defining the grid of values is time-consuming since we need to look at the parameters, read the docs and define the values we want to try. We can automate this process.

Example:

```python
from sklearn_evaluation.grid import RandomForestClassifierGrid

# grid has values: tiny, small, medium, large, xlarge
model = RandomForestClassifierGrid(grid="tiny")

# train all models and return the best one
model.fit()
```

This will essentially be a wrapper of [GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html) and a given model (in this case, [RandomForestClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html)), we can repeat it for other popular models (perhaps XGBoost?)

On top of this, we can provide plotting, evaluation features.

```python
model.confusion_matrix()

model.roc()

model.feature_importances()

# https://sklearn-evaluation.ploomber.io/en/latest/api/plot.html#grid-search
model.grid_search()
```

in version 2, we can add some caching:

```python
model = RandomForestClassifierGrid(grid="tiny")
model.fit()


model.set_grid(grid="small")
# caches results from "tiny", only trains the delta
model.fit()
```

### PR DESCRIPTION
**Pull request by yafimvo**
## Describe your changes
Pre-made RandomForestClassifier added 

## Usage

```python
from sklearn_evaluation.grid import RandomForestClassifierGrid
from sklearn import datasets
from sklearn.model_selection import train_test_split

X, y = datasets.make_classification(200, 20, n_informative=5, class_sep=0.65)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

# grid has values: tiny, small, medium, large, xlarge
model = RandomForestClassifierGrid(grid="large")

# train all models and return the best one
model.fit(X, y)
```

Plot results
```python
model.confusion_matrix()

model.roc()

model.feature_importances()

# https://sklearn-evaluation.ploomber.io/en/latest/api/plot.html#grid-search
model.grid_search()
```
## Issue ticket number and link
Closes #183 

## Checklist before requesting a review
**Not ready for review**

----
:books: Documentation preview :books:: https://sklearn-evaluation--215.org.readthedocs.build/en/215/

### CODE REVIEW

**Comment by yafimvo** [comment]
@edublancas please check if this is the right direction so I can complete this PR with proper tests and documentation.

**Comment by edublancas** [comment]
I almost missed this, please tag me as reviewer, otherwise if you just tag me, there's a higher chance I don't see it.

my feedback:

overall this looks great! What will make things easier for us is if you add a new section in the docs describing how to use this new module, this way we'll be able to click on the doc preview, then click on the binder link and then run your code. to ensure a binder button is added, you need to make your tutorial runnable by adding some metadata (here's an example: https://raw.githubusercontent.com/ploomber/sklearn-evaluation/master/docs/quickstart/classification.md) - the content between `---` is the metadata

I see that plotting methods such as `confusion_matrix` are using the training data for plotting. while this is acceptable, it's common practice to have a separate test set and use this for evaluation/plotting. So we should give that option:

```python
clf = grid.RandomForestClassifierGrid("tiny")
clf.fit(X, y)

clf.set_test_data(X_test, y_test)

# uses X_test and y_test
clf.confusion_matrix()
```

maybe if they don't call `set_test_data` we can use the training data but show a warning? It's usually a bad idea to compute metrics on the training data, but some users might still want to do it (e.g., to evaluate under/overfitting)

also, it should be clear in the docs and tutorial that calling the plotting methods will use the `best_estimator_` - I'm unsure how GridSearchCV defines what "best" is so we should look into that to see how we can expose a setting that allows users to change the definition of "best"

as a convention, `fit` should return `self`. Also, I see that the `GridSearchCV` obj is accessible from `self.classifier`, I think a better name would be `grid_search_cv`, another convention is that attributes that are accessible after calling `fit`, should have a `_` suffix, so let's make it `grid_search_cv_`

> Cool stuff.
Does the user has any control on those? like changing the estimators criterion etc?
It'd probably change frequently

regarding Ido's point. I think this is important. However, I'd say let's push this to v2. Let's deploy this first and then if people like it, we can think of ways for customizing the grids.

**Comment by yafimvo** [comment]
@edublancas  I think we should define the parameters for the different grid sizes to make the test and PR complete. 

Also, I added a new example. Please verify it's a good example and that it's OK to use this dataset.

took it from here
https://github.com/sharmaroshan/Heart-UCI-Dataset

https://www.kaggle.com/datasets/redwankarimsony/heart-disease-data

**Comment by yafimvo** [comment]
@edublancas we need to define grid sizes. now it's just initialized with arbitrary data.

```python

        {
            GridTypes.TINY: {
                "n_estimators": [1, 10],
                "criterion": ["gini", "entropy"],
                "max_features": ["sqrt", "log2"],
            },
            GridTypes.SMALL: {
                "n_estimators": [1, 10, 50],
                "criterion": ["gini", "entropy"],
                "max_features": ["sqrt", "log2"],
            },
            GridTypes.MEDIUM: {
                "n_estimators": [1, 10, 50, 100],
                "criterion": ["gini", "entropy"],
                "max_features": ["sqrt", "log2"],
            },
            GridTypes.LARGE: {
                "n_estimators": [1, 5],
                "max_features": ["sqrt", "log2"],
                "max_depth": [2, 4],
                "min_samples_split": [2, 5],
                "min_samples_leaf": [1, 2],
                "bootstrap": [True],
            },
            GridTypes.X_LARGE: {
                "n_estimators": [1, 10, 50, 100],
                "max_features": ["sqrt", "log2"],
                "max_depth": [2, 4],
                "min_samples_split": [2, 5],
                "min_samples_leaf": [1, 2],
                "bootstrap": [True],
            },
        }

```

File: docs/classification/pre_made_grids.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.14.1
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# Pre made grids
```
Comment (edublancas) [inline docs/classification/pre_made_grids.md:14]: let's change this to "Grid Search", also let's change the name of the file

**Review by edublancas — CHANGES_REQUESTED** [review]
just a minor comment.

Re defining the grids: yeah good point.

we need some research to do here; there aren't any generalizable rules to know which sets of values are appropriate since it highly depends on the data, but we can define some sensible defaults:

for n_estimators, I think a minimum of 50 is all the way up to 500. usually, the higher this parameter is the better (at the cost of increasing runtime), so we have tiny just do: 50, 100, 500, and then as we increase the size of the grid, add more values in the middle

criterion: there aren't many values here so tiny can be `squared_error`, the default, and the others increase the criterion (there are 4 in total), sorted by popularity so we can have the x large try them all

`min_samples_split`: to adjust for different number of features. according to [this](https://towardsdatascience.com/how-to-tune-a-decision-tree-f03721801680), which cites this [paper](https://arxiv.org/abs/1812.02207). the ideal tends to be between 1 and 40. so we can try tiny with [2, 4], then increase the grid all the way up to 40, e.g., xlarge can be [2,4,8,16,24,40]

according to the same source, `min_samples_leaf` tends to be between 1 and 20, we can use the same logic

max_features we can do 1.0 for tiny and add the other two suggested values sqrt2 and log2 in the other grids

we can ignore the other parameters.

**Comment by yafimvo** [comment]
@edublancas 2 things
1. according to the [RandomForestClassifier API](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn-ensemble-randomforestclassifier) the `criterion` takes 3 possible arguments: {`gini`, `entropy`, `log_loss`}. Change to these instead of `squared_error`?
2. Increasing the grid size increases the run time, which affects the tests. How do we want to tackle it? should I test only "tiny" grid?

** I pushed the changes to make it clear. 
1. Fixed namings
2. New grid values
3. Testing only "tiny" grid

**Comment by idomic** [comment]
1. Yes, we should adjust according to the specs.
2. We'd probably want a run for each at least on the first stage IMO (and as long as it doesn't block the CI completely), and then we can convert to tiny. Another option is to cover all of the combinations via tiny and that should be a good enough test.

**Comment by edublancas** [comment]
> according to the [RandomForestClassifier API](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn-ensemble-randomforestclassifier) the criterion takes 3 possible arguments: {gini, entropy, log_loss}. Change to these instead of squared_error?

you're right. I was looking at the regressor docs, squared_error isn't a valid value for the classifier

**Review by edublancas — CHANGES_REQUESTED** [review]
looks good, let's add telemetry and we'll be ready to merge

**Review by edublancas — APPROVED** [review]
